### PR TITLE
Presentation: Ruleset versioning

### PR DESCRIPTION
**Pull request by grigasp**
Introduce `version` attribute for `Ruleset`. Take it into account when embedding rulesets.

Also, _PresentationRules_ model and partition should be marked private.

Related imodel02 PR: https://bentleycs.visualstudio.com/iModelTechnologies/_git/imodel02/pullrequest/167554

### CODE REVIEW

File: presentation/backend/src/presentation-backend/domain/RulesetElements.ts [hunk 1]
```diff
@@ -20,15 +22,19 @@ export class Ruleset extends DefinitionElement {
...
-   * @param iModelDb    - db the ruleset is supposed to be inserted into
+   * @param iModelDb DB the ruleset is supposed to be inserted into
+   * @param modelId ID of a the model this ruleset should be created in
+   * @param ruleset The ruleset code is being created for
    */
-  public static createRulesetCode(modelId: Id64String, rulesetId: string, iModelDb: IModelDb) {
+  public static createRulesetCode(iModelDb: IModelDb, modelId: Id64String, ruleset: PresentationRuleset) {
+    let codeValue = ruleset.id;
+    if (ruleset.version)
+      codeValue += `.${normalizeVersion(ruleset.version)}`;
```
Comment (roluk) [inline presentation/backend/src/presentation-backend/domain/RulesetElements.ts:32]: I think `codeValue` would be easier to scan quickly with the eyes if we used a different separator than what's already present in the version string. How about `@`?

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts [hunk 2]
```diff
@@ -17,16 +18,42 @@ import {
...
+   */
+  skip?: "never" | "same-id" | "same-id-and-version";
+
+  /**
+   * Which existing rulesets should be replaced when we insert a new one.
+   * Defaults to `same-id-and-version`.
+   */
+  replace?: "same-id" | "same-id-older" | "same-id-and-version";
+}
+
```
Comment (roluk) [inline presentation/backend/src/presentation-backend/RulesetEmbedder.ts:56]: I haven't looked too deeply into this, but at first glance it feels like `skip` and `replace` options overlap. What should happen when we specify both as `same-id`?

**Review by roluk — CHANGES_REQUESTED** [review]
Seems ok, but `RulesetInsertOptions` documentation could use some improvement.

File: presentation/backend/src/presentation-backend/domain/RulesetElements.ts (hunk 1 above)
Comment (grigasp) [reply presentation/backend/src/presentation-backend/domain/RulesetElements.ts]: Agree. Changed to `@`.

Comment (roluk) [reply presentation/backend/src/presentation-backend/domain/RulesetElements.ts]: Does this require a change in native side?

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts (hunk 2 above)
Comment (grigasp) [reply presentation/backend/src/presentation-backend/RulesetEmbedder.ts]: Yes, they overlap (e.g. in your given case, if a ruleset with same id is found, then we just skip insertion and don't even use at `replace` value). 

However, there are cases when both are needed, e.g. when we want to skip inserting if there's a ruleset with same id and version, but if it's not - we want to specify what to replace: all rulesets with same id / only older rulesets. And I didn't want to add long values like `skip-on-same-id-and-version-and-replace-all`, `skip-on-same-id-and-version-and-replace-older`, etc..

File: presentation/backend/src/presentation-backend/domain/RulesetElements.ts (hunk 1 above)
Comment (grigasp) [reply presentation/backend/src/presentation-backend/domain/RulesetElements.ts]: No, CodeValue is only used to ensure uniqueness. It's not used used in native code.

**Comment by grigasp** [comment]
> Seems ok, but `RulesetInsertOptions` documentation could use some improvement.

What improvements do you suggest?

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts (hunk 2 above)
Comment (roluk) [reply presentation/backend/src/presentation-backend/RulesetEmbedder.ts]: If only rulesets with the same id are always going to get replaced, how about documenting this fact and renaming options to `"all" | "all-smaller" | "exact"`?

Comment (grigasp) [reply presentation/backend/src/presentation-backend/RulesetEmbedder.ts]: Made suggested change. I chose to rename `replace` to `replaceVersions` to make it clear that all options are based on ruleset version and ID is already accounted for.

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts [hunk 3]
```diff
@@ -17,16 +18,48 @@ import {
...
+   * When should insertion be skipped:
+   * - `same-id` - if iModel already contains a ruleset with the same id and **any** version
+   * - `same-id-and-version-eq` - if iModel already contains a ruleset with same id and version
+   * - `same-id-and-version-gte` - if iModel already contains a ruleset with same id and
+   * version equal to greater than of the inserted ruleset.
+   *
+   * Defaults to `same-id-and-version`.
+   */
+  skip?: "never" | "same-id" | "same-id-and-version-eq" | "same-id-and-version-gte";
+
```
Comment (roluk) [inline presentation/backend/src/presentation-backend/RulesetEmbedder.ts:51]: `"never"` option is not documented.

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts [hunk 4]
```diff
@@ -17,16 +18,48 @@ import {
...
+ */
+export interface RulesetInsertOptions {
+  /**
+   * When should insertion be skipped:
+   * - `same-id` - if iModel already contains a ruleset with the same id and **any** version
+   * - `same-id-and-version-eq` - if iModel already contains a ruleset with same id and version
+   * - `same-id-and-version-gte` - if iModel already contains a ruleset with same id and
+   * version equal to greater than of the inserted ruleset.
+   *
+   * Defaults to `same-id-and-version`.
```
Comment (roluk) [inline presentation/backend/src/presentation-backend/RulesetEmbedder.ts:48]: The default seems to be changed to `same-id-and-version-eq`.

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts [hunk 5]
```diff
@@ -17,16 +18,48 @@ import {
...
+   */
+  skip?: "never" | "same-id" | "same-id-and-version-eq" | "same-id-and-version-gte";
+
+  /**
+   * Which existing versions of rulesets with same id should be replaced when we insert a new one:
+   * - `all` - replace all rulesets with same id.
+   * - `all-lower` - replace rulesets with same id an version lower than the version of inserted ruleset.
+   * - `exact` - replace only the ruleset whose id and version matches the inserted ruleset.
```
Comment (roluk) [inline presentation/backend/src/presentation-backend/RulesetEmbedder.ts:56]: Probably a typo in line 55, `an` -> `and`.

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts (hunk 4 above)
Comment (roluk) [inline presentation/backend/src/presentation-backend/RulesetEmbedder.ts:48]: The default seems to have been changed to `same-id-and-version-eq`.

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts [hunk 6]
```diff
@@ -17,16 +18,48 @@ import {
...
+   * version equal to greater than of the inserted ruleset.
+   *
+   * Defaults to `same-id-and-version`.
+   */
+  skip?: "never" | "same-id" | "same-id-and-version-eq" | "same-id-and-version-gte";
+
+  /**
+   * Which existing versions of rulesets with same id should be replaced when we insert a new one:
+   * - `all` - replace all rulesets with same id.
+   * - `all-lower` - replace rulesets with same id an version lower than the version of inserted ruleset.
```
Comment (roluk) [inline presentation/backend/src/presentation-backend/RulesetEmbedder.ts:55]: Probably a typo in line 55, `an` -> `and`.

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts (hunk 3 above)
Comment (roluk) [inline presentation/backend/src/presentation-backend/RulesetEmbedder.ts:51]: Option `"never"` is not documented.

**Review by roluk — CHANGES_REQUESTED** [review]

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts (hunk 2 above)
Comment (roluk) [reply presentation/backend/src/presentation-backend/RulesetEmbedder.ts]: Thanks

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts (hunk 3 above)
Comment (grigasp) [reply presentation/backend/src/presentation-backend/RulesetEmbedder.ts:51]: isn't it self explanatory? :)

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts (hunk 4 above)
Comment (grigasp) [reply presentation/backend/src/presentation-backend/RulesetEmbedder.ts]: fixed, thanks

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts (hunk 6 above)
Comment (grigasp) [reply presentation/backend/src/presentation-backend/RulesetEmbedder.ts]: fixed

File: presentation/backend/src/presentation-backend/RulesetEmbedder.ts (hunk 3 above)
Comment (roluk) [reply presentation/backend/src/presentation-backend/RulesetEmbedder.ts]: Since `skip` is optional, it's unclear why `"never"` option exists.

Comment (grigasp) [reply presentation/backend/src/presentation-backend/RulesetEmbedder.ts]: Because default is not `never`

Comment (roluk) [reply presentation/backend/src/presentation-backend/RulesetEmbedder.ts]: Oh right.

**Review by roluk — APPROVED** [review]

**Pull request merged**